### DISCUSSION TITLE
Uploading files from client

### DISCUSSION THREAD

**Discussion by mihail727** [nuxt/nuxt#16806] [General]
new method for uploading files with version `v3.0.0-rc.2-27526179.f74a523`
`multer` is used to process `form-data`
`useFileSystemAccess` from `@vueuse` is used to open file dialog

<details>
<summary>Client: <code>@/pages/index.vue</code></summary>

```vue
<script setup lang="ts">
	const onClickUpload = async (e: Event) => {
                const { data, open } = useFileSystemAccess({ dataType: 'Blob' });
		await open();

		const formData = new FormData();
		formData.append('file', data.value);

		await useAsyncData('file', () =>
			$fetch('/api/data/upload?itemType=carousel&fileType=document', {
				method: 'POST',
				body: formData,
			}),
		);
	};
</script>

<template>
  <div>
    <button @click="onClickUpload">Upload</button>
  </div>
</template>
```
</details>

<details>
<summary>Server: <code>@/server/api/data/upload.post.ts</code></summary>

```ts
import { CompatibilityEvent, useQuery, callHandler } from 'h3';
import { uploadService } from '../../services';

export default async (event: CompatibilityEvent): Promise<IServerResponse> => {
	const { itemType, fileType } = useQuery(event) as { itemType: ItemType; fileType: FileType };
	try {
		const handler = uploadService().generateHandler(itemType, fileType);
		// @ts-ignore
		await callHandler(handler, event.req, event.res);
		// @ts-ignore
		return { success: true, response: event.req?.file?.filename };
	} catch (e) {
		return { success: false, reason: e.message };
	}
};
```
</details>

<details>
<summary>Upload service: <code>@/server/services/upload-service.ts</code></summary>

```ts
import multer from 'multer';
import type { Options } from 'multer';
import fs from 'fs';
import { v4 } from 'uuid';

export const uploadService = () => {
	let folderPath = process.env.NODE_ENV == 'development' ? './src/public/' : './public/';
	const { limits: templateLimits }: Options = {
		limits: {
			files: 1,
			fieldNameSize: 400,
			fileSize: 80 * 1024 * 1024,
		},
	};

	const createFolderIfNotExist = (path: string) => {
		if (!fs.existsSync(path)) {
			fs.mkdirSync(path, { recursive: true });
		}
	};

	const { filename }: multer.DiskStorageOptions = {
		filename: (_req, file, cb) => {
			const splittedFileName = file.originalname.split('.');
			let name = splittedFileName[0];
			const type = splittedFileName[1];
			name += '-' + v4().slice(0, 5) + '.' + type;
			cb(null, name);
		},
	};

	const destination = (itemType: ItemType, fileType: FileType) => {
		const { destination }: multer.DiskStorageOptions = {
			destination: async (_req, _file, cb) => {
				folderPath += `${fileType + 's'}/${itemType}`;
				createFolderIfNotExist(folderPath);
				cb(null, folderPath);
			},
		};
		return destination;
	};

	const generateHandler = (itemType: ItemType, fileType: FileType) => {
		try {
			if (
				!fileType ||
				(fileType !== 'document' && fileType !== 'icon' && fileType !== 'image')
			)
				throw new Error('File type не указан или указан неверно.');

			if (
				!itemType ||
				(itemType !== 'assortment' && itemType !== 'carousel' && itemType !== 'catalog')
			)
				throw new Error('Item type не указан или указан неверно.');

			const options: Options = {
				limits: {
					...templateLimits,
				},
				storage: multer.diskStorage({
					filename,
					destination: destination(itemType, fileType),
				}),
			};

			return multer(options).single('file');
		} catch (e) {
			throw e;
		}
	};
	return { generateHandler };
};
```
</details>

### How it works
1. User doing request: `/api/data/upload?itemType=carousel&fileType=document`
2. Server
  2.1 Generating file handler for different `itemType` and `fileType`
  2.2 Calling our generated middleware `handler` with `CallHandler(handler, event.req, event.res)`
  2.3 returning `filename` of new uploaded file

### Additional type
- `@/types/server-response.d.ts`
```ts
declare interface IServerResponse<T = any> {
	success: boolean;
	reason?: string;
	statusCode?: number;
	response?: T;
}
```
- `@/types/type-file.d.ts`
```ts
declare type FileType = 'image' | 'document' | 'icon';
```
- `@/types/type-item.d.ts`
```ts
declare type ItemType = 'carousel' | 'assortment' | 'catalog';
```

### Errors and warnings
all works but with some errors/warnings:
1. warning on the line `await callHandler(handler, event.req, event.res)` of `@/server/api/data/upload.post.ts`
![error](https://i.imgur.com/angyGq3.png)
I don't know how to solve this warning, but it can be ignored by `// @ts-ignore`
2. warning on the line `return { success: true, response: event.req?.file?.filename }` of `@/server/api/data/upload.post.ts`
![error](https://i.imgur.com/oEZ9S4k.png)
Also don't know how to solve this warning, but it can be ignored by `// @ts-ignore`
3. **The most important**: Sometimes in `dev` mode, when working with large files, the process may be interrupted or terminated with an error. This does not happen with smaller files.
![error](https://i.imgur.com/oAgNJSm.png)
But all works if u doing this in `production` mode.

**Comment by xsky13** [comment]
Thank you very much, this helped me a lot

**Comment by lesion** [comment]
thanks! But I did not found a way to serve uploaded files in production:
 `public/` path seems to work in dev but not in production. any clue?

**Reply by mihail727** [reply]
I didn't understand your question... but maybe you should redefine the variable in `@/server/services/upload-service.ts`:
this line: `let folderPath = process.env.NODE_ENV == 'development' ? './src/public/' : './public/';`

**Reply by mihail727** [reply]
I forgot to mention that I use a custom project structure
https://github.com/mihail727/nuxt-template

**Reply by lesion** [reply]
thanks for reply, maybe I'm off topic here I comment it here anyway for posterity.
Upload is working but I cannot serve uploaded files.
eg.

```
<img src='/public/uploaded_image.jpg' />
```
this is working in dev mode but not in production.
probably related to this one: https://github.com/nuxt/framework/issues/2261

nuxt v3.0.0-rc.8-27685975.80abca9

**Reply by mihail727** [reply]
> thanks for reply, maybe I'm off topic here I comment it here anyway for posterity. Upload is working but I cannot serve uploaded files. eg.
> 
> ```
> <img src='/public/uploaded_image.jpg' />
> ```
> 
> this is working in dev mode but not in production. probably related to this one: #2261
> 
> nuxt v3.0.0-rc.8-27685975.80abca9

did you redefine the function that generates the path of the loaded image?
this two:
```ts
const { filename }: multer.DiskStorageOptions = {
		filename: (_req, file, cb) => {
			const splittedFileName = file.originalname.split('.');
			let name = splittedFileName[0];
			const type = splittedFileName[1];
			name += '-' + v4().slice(0, 5) + '.' + type;
			cb(null, name);
		},
	};

	const destination = (itemType: ItemType, fileType: FileType) => {
		const { destination }: multer.DiskStorageOptions = {
			destination: async (_req, _file, cb) => {
				folderPath += `${fileType + 's'}/${itemType}`;
				createFolderIfNotExist(folderPath);
				cb(null, folderPath);
			},
		};
		return destination;
	};
```
after client reuqest like this: `/api/data/upload?itemType=carousel&fileType=image`
this image will be saved in dev mode: `./src/public/images/carousel/image-someHash.webp` in prod mode: `./public/images/carousel/image-someHash.webp`

**Reply by mihail727** [reply]
I'm sorry, my English is not very good and maybe it's just difficult for me to understand the essence of the question.

did I understand correctly? you upload an image to the server and where on the same site is there an image tag that is already trying to link to this image?

**Reply by mihail727** [reply]
I wonder if you are right about the discussion above....
I just uploaded these pictures using a self-written image processing service
`imageMobileLazyFilePath.value = '/api/data/file?fileType=image&path=\images\carousel\mx2-3fged.webp&w=520.3&b=50'`

**Reply by mihail727** [reply]
```ts
import fs from 'fs';
import gm from 'gm';

export const imageService = async (
	imagePath: string, // Исходный файл
	width: number,
	quality: number = 80,
	format: 'webp' | 'jpg' | 'png' = 'png',
	blur: number = 0,
) => {
	if (isNaN(width)) throw new Error('Ширина должна быть указана в числовом формате.');
	if (isNaN(quality)) quality = 100;
	if (isNaN(blur)) blur = 0;
	if (!format || (format != 'jpg' && format != 'webp' && format != 'png')) format = 'png';

	const clearImageName = getClearName(imagePath); // loader
	const clearFolderPath = getClearFolderPath(imagePath, clearImageName); // ./src/public/images/carousel/

	const searchImageName = clearImageName + `-[${width}_${quality}_${blur}].${format}`; // loader-[100_80_30].webp
	const searchImageWithPath = clearFolderPath + searchImageName; // ./src/public/images/carousel/loader-[100_80_30].webp

	if (!isFileExist(searchImageWithPath)) {
		await new Promise((resolve) => {
			let image = gm(imagePath).resize(width).setFormat(format).quality(quality);
			if (blur !== 0) image.blur(blur, 30);

			image.write(searchImageWithPath, (e) => {
				if (e) throw e;
				resolve(0);
			});
		});
	}
	return searchImageWithPath;
};

const getClearFolderPath = (path: string, clearImageName: string) => {
	const folderWithName = '.' + path.split('.')[1];
	return folderWithName.split(clearImageName)[0];
};

const getClearName = (path: string) => {
	const folderWithName = '.' + path.split('.')[1];
	return folderWithName.split('/')[folderWithName.split('/').length - 1];
};

const isFileExist = (path: string) => {
	try {
		fs.accessSync(path);
		return true;
	} catch (e) {
		return false;
	}
};
```

**Reply by lesion** [reply]
mine is simpler:
```javascript
  const { filename }: multer.DiskStorageOptions = {
    filename: (_req, file, cb) => {
      const type = path.extname(file.originalname)
      cb(null, v4() + type)
    }
  }
```
this is working. files are uploaded and written to disk:

```bash
➜  .output git:(main) ✗ ls -la public 
total 160
drwxr-xr-x 3 les les  4096 Aug 22 15:46 .
drwxr-xr-x 4 les les  4096 Aug 22 15:20 ..
-rw-r--r-- 1 les les 72137 Aug 22 15:20 3eca088e-d477-4d0e-9820-a1cd36ec9d6f.png
-rw-r--r-- 1 les les 72137 Aug 22 15:19 f5ac1a17-48f5-42e9-848d-9b97f092370c.png
-rw-r--r-- 1 les les  2444 Aug 22 15:19 manifest.json
drwxr-xr-x 2 les les  4096 Aug 22 15:22 _nuxt
```

visiting `http://localhost:3000/public/3eca088e-d477-4d0e-9820-a1cd36ec9d6f.png` or without public returns me a 404 (only in production, `yarn build` and `yarn preview`)

**Reply by GamingWolf** [reply]
Did you ever find a fix for this? I am facing the same issue. File uploads just fine, but the Vue router complains that it does not exist.

**Reply by lesion** [reply]
I serve images via API:

https://git.lattuga.net/lesion/cosette/src/main/server/api/image/[uuid].js

https://git.lattuga.net/lesion/cosette/src/main/components/cosetta.vue#L11

**Reply by GamingWolf** [reply]
Yeah, that is the solution I settled on as well. Had just hoped that maybe you found a fix for the Vue Router being mad.
